File: rdf-infographic-skill/scripts/validate-infographic.py
#!/usr/bin/env python3
"""
HTML Infographic Validation Script

Validates generated HTML infographics against quality checklist.

Usage:
    python3 validate-infographic.py output.html
"""

import sys
import re
from pathlib import Path
from typing import Dict, List, Tuple


class InfographicValidator:
    """Validate generated infographics"""
    
    def __init__(self, html_path: str):
        """Initialize validator"""
        self.html_path = html_path
        try:
            with open(html_path, 'r', encoding='utf-8') as f:
                self.html = f.read()
        except Exception as e:
            print(f"Error reading file: {e}", file=sys.stderr)
            sys.exit(1)
        
        self.results = {
            'passed': [],
            'failed': [],
            'warnings': []
        }
    
    def validate(self) -> bool:
        """Run all validation checks"""
        print("🔍 Validating infographic...\n")
        
        self._check_navigation()
        self._check_content()
        self._check_interactions()
        self._check_metadata()
        self._check_accessibility()
        self._check_performance()
        
        return self._print_results()
    
    def _check_navigation(self):
        """Check navigation panel implementation"""
        section = "Navigation Control"
        
        # Check for floating nav
        if 'floatingNav' in self.html or 'float-nav' in self.html.lower():
            self.results['passed'].append(f"{section}: Found navigation panel")
        else:
            self.results['failed'].append(f"{section}: Navigation panel not found")
        
        # Check for draggable
        if 'mousedown' in self.html and 'mousemove' in self.html:
            self.results['passed'].append(f"{section}: Drag functionality implemented")
        else:
            self.results['warnings'].append(f"{section}: Drag functionality not found")
        
        # Check for close/open logic
        if 'isManuallyClosed' in self.html or 'close' in self.html.lower():
            self.results['passed'].append(f"{section}: Close/open logic found")
        else:
            self.results['warnings'].append(f"{section}: Close/open logic not obvious")
        
        # Check for inactivity timer
        if 'inactivity' in self.html.lower() or '10000' in self.html:
            self.results['passed'].append(f"{section}: Inactivity timer implemented")
        else:
            self.results['warnings'].append(f"{section}: Inactivity timer not found")
        
        # Check for pin marker
        if 'pin' in self.html.lower():
            self.results['passed'].append(f"{section}: Pin marker restoration found")
        else:
            self.results['warnings'].append(f"{section}: Pin marker not obvious")
    
    def _check_content(self):
        """Check content structure"""
        section = "Content Structure"
        
        # Check for multiple entity types
        entity_types = len(re.findall(r'data-entity-type|entity-type|type:', self.html, re.IGNORECASE))
        if entity_types > 3:
            self.results['passed'].append(f"{section}: Multiple entity types represented ({entity_types}+)")
        else:
            self.results['warnings'].append(f"{section}: Limited entity types found")
        
        # Check for acronyms
        if 'acronym' in self.html.lower() or 'abbr' in self.html.lower():
            self.results['passed'].append(f"{section}: Acronym expansion found")
        else:
            self.results['warnings'].append(f"{section}: Acronym expansion not found")
        
        # Check for links
        links = len(re.findall(r'<a[^>]*href=', self.html))
        if links > 5:
            self.results['passed'].append(f"{section}: Entity links implemented ({links} links)")
        else:
            self.results['warnings'].append(f"{section}: Limited links found ({links})")
        
        # Check for section IDs
        section_ids = len(re.findall(r'id="[^"]*"', self.html))
        if section_ids > 5:
            self.results['passed'].append(f"{section}: Section anchors implemented")
        else:
            self.results['warnings'].append(f"{section}: Limited section anchors")
        
        # Check for problem-solution
        if ('problem' in self.html.lower() and 'solution' in self.html.lower()) or \
           ('challenge' in self.html.lower() and 'overcome' in self.html.lower()):
            self.results['passed'].append(f"{section}: Problem-solution framing found")
        else:
            self.results['warnings'].append(f"{section}: Problem-solution framing not obvious")
    
    def _check_interactions(self):
        """Check interactive elements"""
        section = "Interactive Elements"
        
        # Check for accordion
        if 'accordion' in self.html.lower():
            self.results['passed'].append(f"{section}: Accordion component found")
        else:
            self.results['warnings'].append(f"{section}: Accordion not found")
        
        # Check for smooth transitions
        if 'transition' in self.html.lower() or 'cubic-bezier' in self.html:
            self.results['passed'].append(f"{section}: Smooth transitions implemented")
        else:
            self.results['warnings'].append(f"{section}: Smooth transitions not obvious")
        
        # Check for hover effects
        if ':hover' in self.html or 'hover:' in self.html:
            self.results['passed'].append(f"{section}: Hover effects implemented")
        else:
            self.results['warnings'].append(f"{section}: Hover effects not obvious")
        
        # Check for scroll animations
        if 'IntersectionObserver' in self.html or 'intersection' in self.html.lower():
            self.results['passed'].append(f"{section}: Scroll animations with Intersection Observer")
        elif 'scroll' in self.html.lower():
            self.results['warnings'].append(f"{section}: Scroll handling found (may not use Intersection Observer)")
        else:
            self.results['warnings'].append(f"{section}: Scroll animations not found")
        
        # Check for lightbox
        if 'lightbox' in self.html.lower() or 'modal' in self.html.lower():
            self.results['passed'].append(f"{section}: Image lightbox found")
        else:
            self.results['warnings'].append(f"{section}: Image lightbox not found")
        
        # Check responsive design
        if 'media query' in self.html.lower() or '@media' in self.html:
            self.results['passed'].append(f"{section}: Media queries for responsiveness")
        else:
            self.results['warnings'].append(f"{section}: Media queries not obvious")
    
    def _check_metadata(self):
        """Check metadata and SEO"""
        section = "Metadata & SEO"
        
        # Check JSON-LD
        if 'application/ld+json' in self.html:
            self.results['passed'].append(f"{section}: JSON-LD structured data present")
        else:
            self.results['failed'].append(f"{section}: JSON-LD structured data missing")
        
        # Check microdata
        if 'itemscope' in self.html or 'itemprop' in self.html:
            self.results['passed'].append(f"{section}: Microdata annotations present")
        else:
            self.results['warnings'].append(f"{section}: Microdata annotations not found")
        
        # Check Open Graph
        if 'og:' in self.html or 'property="og' in self.html:
            self.results['passed'].append(f"{section}: Open Graph tags present")
        else:
            self.results['warnings'].append(f"{section}: Open Graph tags not found")
        
        # Check meta tags
        meta_tags = len(re.findall(r'<meta[^>]*>', self.html))
        if meta_tags > 5:
            self.results['passed'].append(f"{section}: Comprehensive meta tags ({meta_tags})")
        else:
            self.results['warnings'].append(f"{section}: Limited meta tags ({meta_tags})")
        
        # Check heading hierarchy
        h1_count = len(re.findall(r'<h1[^>]*>', self.html))
        h2_count = len(re.findall(r'<h2[^>]*>', self.html))
        if h1_count == 1 and h2_count > 2:
            self.results['passed'].append(f"{section}: Proper heading hierarchy")
        elif h1_count != 1:
            self.results['failed'].append(f"{section}: Page should have exactly 1 H1 (found {h1_count})")
        else:
            self.results['warnings'].append(f"{section}: Limited H2 tags")
        
        # Check canonical URL
        if 'canonical' in self.html.lower():
            self.results['passed'].append(f"{section}: Canonical URL present")
        else:
            self.results['warnings'].append(f"{section}: Canonical URL not found")
    
    def _check_accessibility(self):
        """Check accessibility features"""
        section = "Accessibility"
        
        # Check alt text
        images = len(re.findall(r'<img[^>]*>', self.html))
        alt_text = len(re.findall(r'<img[^>]*alt="[^"]*"', self.html))
        if alt_text > 0:
            self.results['passed'].append(f"{section}: Alt text on images ({alt_text}/{images})")
        else:
            self.results['warnings'].append(f"{section}: Alt text on images not found")
        
        # Check for focus states
        if 'focus' in self.html.lower():
            self.results['passed'].append(f"{section}: Focus states implemented")
        else:
            self.results['warnings'].append(f"{section}: Focus states not obvious")
        
        # Check language attribute
        if 'lang=' in self.html:
            self.results['passed'].append(f"{section}: Language attribute set")
        else:
            self.results['warnings'].append(f"{section}: Language attribute not found")
        
        # Check for semantic HTML
        semantic_tags = len(re.findall(r'<(article|section|nav|header|footer)[^>]*>', self.html))
        if semantic_tags > 3:
            self.results['passed'].append(f"{section}: Semantic HTML5 tags ({semantic_tags})")
        else:
            self.results['warnings'].append(f"{section}: Limited semantic HTML tags")
        
        # Check for color contrast info (in comments or CSS)
        if 'contrast' in self.html.lower() or '4.5' in self.html or '5.3' in self.html:
            self.results['passed'].append(f"{section}: Color contrast considerations documented")
        else:
            self.results['warnings'].append(f"{section}: Color contrast not documented")
    
    def _check_performance(self):
        """Check performance considerations"""
        section = "Performance"
        
        # Check file size
        file_size_kb = len(self.html) / 1024
        if file_size_kb < 500:
            self.results['passed'].append(f"{section}: Reasonable file size ({file_size_kb:.1f} KB)")
        elif file_size_kb < 1000:
            self.results['warnings'].append(f"{section}: Large file size ({file_size_kb:.1f} KB)")
        else:
            self.results['failed'].append(f"{section}: Very large file size ({file_size_kb:.1f} KB)")
        
        # Check for external resources from CDN
        cdn_resources = len(re.findall(r'https://(cdn\.|[a-z]+\.com/)', self.html))
        if cdn_resources > 0:
            self.results['passed'].append(f"{section}: Using CDN resources ({cdn_resources})")
        else:
            self.results['warnings'].append(f"{section}: Not using CDN resources")
        
        # Check for CSS animations vs JavaScript
        if '@keyframes' in self.html and 'animation:' in self.html:
            self.results['passed'].append(f"{section}: CSS animations used for performance")
        else:
            self.results['warnings'].append(f"{section}: CSS animations not obvious")
        
        # Check for Intersection Observer (performant scrolling)
        if 'IntersectionObserver' in self.html:
            self.results['passed'].append(f"{section}: Intersection Observer for efficient scrolling")
        else:
            self.results['warnings'].append(f"{section}: Scroll handling may not be optimized")
        
        # Check for lazy loading
        if 'lazy' in self.html.lower() or 'data-src' in self.html:
            self.results['passed'].append(f"{section}: Lazy loading implemented")
        else:
            self.results['warnings'].append(f"{section}: Lazy loading not found")
    
    def _print_results(self) -> bool:
        """Print validation results"""
        total = len(self.results['passed']) + len(self.results['failed'])
        passed = len(self.results['passed'])
        
        print("✅ PASSED")
        for item in self.results['passed']:
            print(f"  • {item}")
        
        if self.results['warnings']:
            print("\n⚠️  WARNINGS")
            for item in self.results['warnings']:
                print(f"  • {item}")
        
        if self.results['failed']:
            print("\n❌ FAILED")
            for item in self.results['failed']:
                print(f"  • {item}")
        
        print(f"\n📊 Score: {passed}/{total} checks passed ({100*passed//total}%)")
        
        return len(self.results['failed']) == 0


def main():
    """CLI interface"""
    if len(sys.argv) < 2:
        print("Usage: python3 validate-infographic.py <html_file>")
        sys.exit(1)
    
    html_file = sys.argv[1]
    validator = InfographicValidator(html_file)
    success = validator.validate()
    sys.exit(0 if success else 1)


if __name__ == '__main__':
    main()

File: rdf-infographic-skill/scripts/validate-infographic.ts
/**
 * HTML Infographic Validation Script — TypeScript edition (Node.js ≥ 18, no npm deps).
 * Identical behavior to validate-infographic.py.
 *
 * Usage:
 *   npx tsx validate-infographic.ts output.html
 */

import { readFileSync } from "node:fs";

interface Results {
  passed: string[];
  failed: string[];
  warnings: string[];
}

class InfographicValidator {
  private readonly html: string;
  private readonly results: Results = { passed: [], failed: [], warnings: [] };

  constructor(htmlPath: string) {
    try {
      this.html = readFileSync(htmlPath, "utf-8");
    } catch (err) {
      process.stderr.write(`Error reading file: ${(err as Error).message}\n`);
      process.exit(1);
    }
  }

  validate(): boolean {
    console.log("Validating infographic...\n");
    this.checkNavigation();
    this.checkContent();
    this.checkInteractions();
    this.checkMetadata();
    this.checkAccessibility();
    this.checkPerformance();
    return this.printResults();
  }

  private pass(msg: string): void { this.results.passed.push(msg); }
  private fail(msg: string): void { this.results.failed.push(msg); }
  private warn(msg: string): void { this.results.warnings.push(msg); }

  private checkNavigation(): void {
    const s = "Navigation Control";
    const html = this.html;

    if (html.includes("floatingNav") || html.toLowerCase().includes("float-nav"))
      this.pass(`${s}: Found navigation panel`);
    else this.fail(`${s}: Navigation panel not found`);

    if (html.includes("mousedown") && html.includes("mousemove"))
      this.pass(`${s}: Drag functionality implemented`);
    else this.warn(`${s}: Drag functionality not found`);

    if (html.includes("isManuallyClosed") || html.toLowerCase().includes("close"))
      this.pass(`${s}: Close/open logic found`);
    else this.warn(`${s}: Close/open logic not obvious`);

    if (html.toLowerCase().includes("inactivity") || html.includes("10000"))
      this.pass(`${s}: Inactivity timer implemented`);
    else this.warn(`${s}: Inactivity timer not found`);

    if (html.toLowerCase().includes("pin"))
      this.pass(`${s}: Pin marker restoration found`);
    else this.warn(`${s}: Pin marker not obvious`);
  }

  private checkContent(): void {
    const s = "Content Structure";
    const html = this.html;

    const entityTypes = (html.match(/data-entity-type|entity-type|type:/gi) ?? []).length;
    if (entityTypes > 3) this.pass(`${s}: Multiple entity types represented (${entityTypes}+)`);
    else this.warn(`${s}: Limited entity types found`);

    if (html.toLowerCase().includes("acronym") || html.toLowerCase().includes("abbr"))
      this.pass(`${s}: Acronym expansion found`);
    else this.warn(`${s}: Acronym expansion not found`);

    const links = (html.match(/<a[^>]*href=/g) ?? []).length;
    if (links > 5) this.pass(`${s}: Entity links implemented (${links} links)`);
    else this.warn(`${s}: Limited links found (${links})`);

    const sectionIds = (html.match(/id="[^"]*"/g) ?? []).length;
    if (sectionIds > 5) this.pass(`${s}: Section anchors implemented`);
    else this.warn(`${s}: Limited section anchors`);

    const lc = html.toLowerCase();
    if ((lc.includes("problem") && lc.includes("solution")) ||
        (lc.includes("challenge") && lc.includes("overcome")))
      this.pass(`${s}: Problem-solution framing found`);
    else this.warn(`${s}: Problem-solution framing not obvious`);
  }

  private checkInteractions(): void {
    const s = "Interactive Elements";
    const html = this.html;
    const lc = html.toLowerCase();

    if (lc.includes("accordion")) this.pass(`${s}: Accordion component found`);
    else this.warn(`${s}: Accordion not found`);

    if (lc.includes("transition") || html.includes("cubic-bezier"))
      this.pass(`${s}: Smooth transitions implemented`);
    else this.warn(`${s}: Smooth transitions not obvious`);

    if (html.includes(":hover") || html.includes("hover:"))
      this.pass(`${s}: Hover effects implemented`);
    else this.warn(`${s}: Hover effects not obvious`);

    if (html.includes("IntersectionObserver") || lc.includes("intersection"))
      this.pass(`${s}: Scroll animations with Intersection Observer`);
    else if (lc.includes("scroll"))
      this.warn(`${s}: Scroll handling found (may not use Intersection Observer)`);
    else this.warn(`${s}: Scroll animations not found`);

    if (lc.includes("lightbox") || lc.includes("modal"))
      this.pass(`${s}: Image lightbox found`);
    else this.warn(`${s}: Image lightbox not found`);

    if (lc.includes("media query") || html.includes("@media"))
      this.pass(`${s}: Media queries for responsiveness`);
    else this.warn(`${s}: Media queries not obvious`);
  }

  private checkMetadata(): void {
    const s = "Metadata & SEO";
    const html = this.html;

    if (html.includes("application/ld+json"))
      this.pass(`${s}: JSON-LD structured data present`);
    else this.fail(`${s}: JSON-LD structured data missing`);

    if (html.includes("itemscope") || html.includes("itemprop"))
      this.pass(`${s}: Microdata annotations present`);
    else this.warn(`${s}: Microdata annotations not found`);

    if (html.includes("og:") || html.includes('property="og'))
      this.pass(`${s}: Open Graph tags present`);
    else this.warn(`${s}: Open Graph tags not found`);

    const metaTags = (html.match(/<meta[^>]*>/g) ?? []).length;
    if (metaTags > 5) this.pass(`${s}: Comprehensive meta tags (${metaTags})`);
    else this.warn(`${s}: Limited meta tags (${metaTags})`);

    const h1Count = (html.match(/<h1[^>]*>/g) ?? []).length;
    const h2Count = (html.match(/<h2[^>]*>/g) ?? []).length;
    if (h1Count === 1 && h2Count > 2) this.pass(`${s}: Proper heading hierarchy`);
    else if (h1Count !== 1) this.fail(`${s}: Page should have exactly 1 H1 (found ${h1Count})`);
    else this.warn(`${s}: Limited H2 tags`);

    if (html.toLowerCase().includes("canonical"))
      this.pass(`${s}: Canonical URL present`);
    else this.warn(`${s}: Canonical URL not found`);
  }

  private checkAccessibility(): void {
    const s = "Accessibility";
    const html = this.html;

    const images = (html.match(/<img[^>]*>/g) ?? []).length;
    const altText = (html.match(/<img[^>]*alt="[^"]*"/g) ?? []).length;
    if (altText > 0) this.pass(`${s}: Alt text on images (${altText}/${images})`);
    else this.warn(`${s}: Alt text on images not found`);

    if (html.toLowerCase().includes("focus"))
      this.pass(`${s}: Focus states implemented`);
    else this.warn(`${s}: Focus states not obvious`);

    if (html.includes("lang="))
      this.pass(`${s}: Language attribute set`);
    else this.warn(`${s}: Language attribute not found`);

    const semanticTags = (html.match(/<(article|section|nav|header|footer)[^>]*>/g) ?? []).length;
    if (semanticTags > 3) this.pass(`${s}: Semantic HTML5 tags (${semanticTags})`);
    else this.warn(`${s}: Limited semantic HTML tags`);

    if (html.toLowerCase().includes("contrast") || html.includes("4.5") || html.includes("5.3"))
      this.pass(`${s}: Color contrast considerations documented`);
    else this.warn(`${s}: Color contrast not documented`);
  }

  private checkPerformance(): void {
    const s = "Performance";
    const html = this.html;

    const fileSizeKb = html.length / 1024;
    if (fileSizeKb < 500) this.pass(`${s}: Reasonable file size (${fileSizeKb.toFixed(1)} KB)`);
    else if (fileSizeKb < 1000) this.warn(`${s}: Large file size (${fileSizeKb.toFixed(1)} KB)`);
    else this.fail(`${s}: Very large file size (${fileSizeKb.toFixed(1)} KB)`);

    const cdnResources = (html.match(/https:\/\/(cdn\.|[a-z]+\.com\/)/g) ?? []).length;
    if (cdnResources > 0) this.pass(`${s}: Using CDN resources (${cdnResources})`);
    else this.warn(`${s}: Not using CDN resources`);

    if (html.includes("@keyframes") && html.includes("animation:"))
      this.pass(`${s}: CSS animations used for performance`);
    else this.warn(`${s}: CSS animations not obvious`);

    if (html.includes("IntersectionObserver"))
      this.pass(`${s}: Intersection Observer for efficient scrolling`);
    else this.warn(`${s}: Scroll handling may not be optimized`);

    if (html.toLowerCase().includes("lazy") || html.includes("data-src"))
      this.pass(`${s}: Lazy loading implemented`);
    else this.warn(`${s}: Lazy loading not found`);
  }

  private printResults(): boolean {
    const total = this.results.passed.length + this.results.failed.length;
    const passed = this.results.passed.length;

    console.log("PASSED");
    for (const item of this.results.passed) console.log(`  - ${item}`);

    if (this.results.warnings.length) {
      console.log("\nWARNINGS");
      for (const item of this.results.warnings) console.log(`  - ${item}`);
    }

    if (this.results.failed.length) {
      console.log("\nFAILED");
      for (const item of this.results.failed) console.log(`  - ${item}`);
    }

    const pct = total > 0 ? Math.floor((100 * passed) / total) : 0;
    console.log(`\nScore: ${passed}/${total} checks passed (${pct}%)`);
    return this.results.failed.length === 0;
  }
}

function main(): void {
  const argv = process.argv.slice(2);
  if (!argv.length) {
    process.stderr.write("Usage: npx tsx validate-infographic.ts <html_file>\n");
    process.exit(1);
  }
  const validator = new InfographicValidator(argv[0]);
  const success = validator.validate();
  process.exit(success ? 0 : 1);
}

main();
